keps/sig-node/20190425-node-readiness-gates.md

```diff
@@ -0,0 +1,390 @@
+actual readiness. It is not straightforward to introduce new conditions into the policy.
+* External systems (such as networking by the cloud provider) must often be configured before pods should be scheduled on a node.
+* Cluster Autoscaler incorrectly assumes a GPU node to be “ready” prior to device plugins and/or device drivers
+being installed which leads to spurious node additions and deletions thereby leading to custom logic in the autoscaler.
+* Logs & metrics are lost if logging & monitoring agents are not online which can lead to sub-optimal (often unexpected)
+user behavior.
+* A node with offline kube-proxy can result in service outage in the event of an application update (via deployments for example).
+* It is difficult for Node Lifecycle Controller and/or node repair systems to detect and remedy issues with extended system
+components of the node since they lack a set of generic & easy to identify signals from those individual components on the node.
+
+As Kubernetes becomes more and more customizable (or extensible) and gets deployed in mission critical systems,
+it is imperative to provide fail safe mechanisms that are built into the system.
+
+The remainder of this proposal introduces an approach for addressing the issue of node health extensibility using
+node readiness gates.
+
+### Goals
+
+* introduce a mechanism that allows for customizable/extensible node readiness checks.
+
+### Non-Goals
+
+* improving time-to-readiness of nodes
+
+## Proposal
+
+Any solution that is chosen for addressing the set of problems stated above should meet the following design requirements:
+
+* Be backwards compatible
+* Be extensible to accommodate variations in kubernetes environments - the set of system microservices can vary across k8s deployments.
+* Ideally, provide a single signal (not a policy) to cluster level components.
+* Ideally, avoid introducing new API primitives.
+* Support self-bootstrapping - scheduling system components (including the kubelet) even if the node is not logically ready.
+
+## Design Details
+
+### Determining Readiness via Pod Selectors
+
+In this design, the indication of node readiness can be expressed by the readiness/health of specific pods running on that node. This approach
+would leverage existing mechnanisms and well-known concepts to determine if a node is ready to receive pods.
+
+A readiness taint will be used to track the overall readiness of a node. The kubelet, node lifecycle controller, other external controllers will
+manage that taint.
+
+The Node object will be extended to include a list of readiness gates like so:
+
+```go
+type NodeSpec struct {
+  ...
+  // Defines the list of readiness gates for a node
+  // +optional
+  ReadinessGates []NodeReadinessGate `json:"readinessGates,omitempty" protobuf:"bytes,7,opt,name=readinessGates"`
+}
+
+// NodeReadinessGate indicates a set of pods that must be ready on a given node before the
```

Comment:
Since the selector may select more than one pod, what is the expected behavior if not all of those selected pods are ready?
Is the gate ready if *at least one* of the selected pods is ready?
Or if *all* of the selected pods are ready?

If it's *all*, then this requires first calculating the set of system-critical daemonset pods that are expected to be scheduled on the node, before trying to evaluate the node readiness. And kubelet can't do that, only the scheduler can. So the *all* semantics would require the cooperation of the scheduler to determine node readiness.

In any case, this has to be clarified here.